ls1intum/Artemis
# `Programming exercises`: Fix check for already existing exercise with same name in GitLab setups

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises
- [ ] ~~I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).~~ not applicable

### Motivation and Context
One instructor tried to create an exercise with short name `ABC`. This did not work since Artemis complained that a project with the same name already exists in GitLab.
This was not the case, only another exercise in the same course with name `ACDEF` already existed.

### Description
The previous solution marked a project as existing if any substring of the project name matched. E.g., a project `COURSE01PROJECTAB` exists. Now the user wants to create a new exercise with short identifier `COURSE01PROJECTA`. This was not possible since the search for this identifier also returned search results matching the different name since they share a common substring in the name.
The new check checks for the _exact_ project key to check if the repositories for this key already exist in GitLab.

As a side-effect this also improves performance of this check:
The previous API call loaded _all_ projects in GitLab that contained the short name _anywhere in their name_. The actual contents of the list was ignored and it was only checked if the list is empty.
The new check limits the search to a single group and also uses a paginated response to limit the number of projects loaded from the GitLab API.

### Steps for Testing

Prerequisites:
- Server with GitLab/Jenkins setup
- 1 Instructor
- 1 Programming Exercise

1. On the details page for the existing exercise: Run the Consistency check (transparent/blue button). This should report no inconsistencies.
2. Note the short name for this exercise (listed under exercise details on the same page).
3. Create a new programming exercise in the same course with a short name that is a substring of the existing one.
4. This should work.

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage





| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| GitLabService.java | 78% | ✅ (100% line/branch for changed code)  |
